File: Examples/WindowsDesktop/MainForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using UnderAutomation.UniversalRobots;

public partial class MainForm : Form
{
    // UR instance that handles connection to the robot
    private readonly UR _ur = new UR();

    public static MainForm Instance { get; private set; }

    #region Initialisation
    public MainForm()
    {
        Instance = this;

        // Catch all unhandled exceptions
        AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
        Application.ThreadException += Application_ThreadException;

        // Load configuration file
        Config.Load();

        InitializeComponent();

        // Add nodes in left menu and instanciate associated control 
        AddNode(new ConnectControl(_ur));
        AddNode(new PrimaryInterfaceControl(_ur));
        AddNode(new ProgramControl(_ur));
        AddNode(new RtdeControl(_ur));
        AddNode(new DashboardControl(_ur));
        AddNode(new XmlRpcControl(_ur));
        AddNode(new SocketControl(_ur));
        AddNode(new SftpControl(_ur));
        AddNode(new SshControl(_ur));
        AddNode(new ToolsControl());
        AddNode(new ArchiveControl(_ur));
        AddNode(new LicenseControl());
        AddNode(new LogControl());

        _ur.InternalErrorOccured += _ur_InternalErrorOccured;

        // Select first node at startup
        SelectNode(leftTreeView.Nodes[0]);

    }


    private void AddNode(IUserControl control)
    {
        var node = leftTreeView.Nodes.Add(control.Title);
        node.Tag = control;
    }
    #endregion

    #region Unhandled exception
    private void Application_ThreadException(object sender, System.Threading.ThreadExceptionEventArgs e)
    {
        CatchApplicationException(e.Exception);
    }

    private void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
    {
        CatchApplicationException(e.ExceptionObject as Exception);
    }

    private void CatchApplicationException(Exception e)
    {
        if (this.InvokeRequired)
        {
            this.Invoke(new Action(() => CatchApplicationException(e)));
            return;
        }

        MessageBox.Show(e?.Message, "An error occurred", MessageBoxButtons.OK, MessageBoxIcon.Error);
    }
    #endregion

    #region User interaction
    private void leftTreeView_NodeMouseClick(object sender, TreeNodeMouseClickEventArgs e)
    {
        SelectNode(e.Node);
    }

    internal T SelectNode<T>() where T : class, IUserControl
    {
        var node = Instance.leftTreeView.Nodes.OfType<TreeNode>().FirstOrDefault(n => n.Tag is T);
        var control = node.Tag as T;
        SelectNode(node);
        return control;
    }

    // Open right control associated to a node
    internal void SelectNode(TreeNode node)
    {
        if (leftTreeView.SelectedNode != node) leftTreeView.SelectedNode = node;

        if (node is null) return;

        mainPanel.SuspendLayout();

        mainPanel.Controls.OfType<IUserControl>().FirstOrDefault()?.OnClose();

        mainPanel.Controls.Clear();

        var control = node.Tag as Control;

        if (control == null)
        {
            panelTitle.Text = "";
            return;
        }

        panelTitle.Text = (node.Tag as IUserControl)?.Title;

        mainPanel.Controls.Add(control);

        control.Dock = DockStyle.Fill;

        mainPanel.ResumeLayout();

        mainPanel.Controls.OfType<IUserControl>().FirstOrDefault()?.OnOpen();
        mainPanel.Controls.OfType<IUserControl>().FirstOrDefault()?.PeriodicUpdate();
    }

    // Open browser to documentation page
    private void lblLink_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
    {
        System.Diagnostics.Process.Start("https://underautomation.com/universal-robots/documentation?f");
    }
    #endregion

    // Timer callback to refresh the control on the right and the left menu
    private void tmrPeriodicUpdate_Tick(object sender, EventArgs e)
    {
        try
        {
            // refresh control on the right
            mainPanel.Controls.OfType<IUserControl>().FirstOrDefault()?.PeriodicUpdate();
        }
        catch { }

        lblStatus.Text = _ur.Enabled ? $"Connected to {_ur.IP}" : "Disconnected";

        try
        {
            // Refresh left menu icons and colors
            foreach (TreeNode node in leftTreeView.Nodes)
            {
                try
                {
                    var ctrl = node?.Tag as IUserControl;
                    if (ctrl == null) continue;

                    var enabled = ctrl.FeatureEnabled;

                    var key = enabled ? "ON" : "OFF";

                    if (key != node.ImageKey)
                    {
                        node.ImageKey = key;
                        node.SelectedImageKey = key;
                        node.ForeColor = enabled ? Color.Green : Color.DarkGray;
                    }
                }
                catch { }
            }
        }
        catch { }
    }

    private void titlePictureBox_MouseDoubleClick(object sender, MouseEventArgs e)
    {
        ScreenshotsGenerator.Generate(this);
    }

    public static void Decompile(string fullName)
    {
        var control = Instance.SelectNode<ArchiveControl>();
        control?.Decompile(fullName);
    }

    public static void InvokeOnMainForm(Delegate action)
    {
        if (Instance.InvokeRequired)
        {
            Instance.Invoke(action);
            return;
        }
        action?.DynamicInvoke();
    }

    #region Show errors
    public readonly List<string> Errors = new List<string>();

    private void lblErrors_Click(object sender, EventArgs e)
    {
        SelectNode<LogControl>();
    }

    // Hide red label after some times
    private void tmrError_Tick(object sender, EventArgs e)
    {
        tmrError.Enabled = false;
        lblErrors.Text = "";
        lblErrors.Visible = false;
    }

    // Callback when an error occured in SDK
    private void _ur_InternalErrorOccured(object sender, InternalErrorEventArgs e)
    {
        var error = e.ToString();
        Errors.Add(error);
        ShowError(error);
    }

    // Display error
    private void ShowError(string error)
    {
        if (InvokeRequired)
        {
            Invoke(new Action(() => ShowError(error)));
            return;
        }

        if (string.IsNullOrEmpty(error)) return;

        if (error.Length > 100) error = error.Substring(0, 100) + " ...";

        lblErrors.Text = error;
        lblErrors.Visible = true;

        // Show label some times
        tmrError.Enabled = true;
    }
    #endregion
}

File: UnderAutomation.UniversalRobots.Showcase.Forms.vb/MainForm.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports UnderAutomation.UniversalRobots
Imports UnderAutomation.UniversalRobots.Common

Public Partial Class MainForm
    Inherits Form

    Private Shared _Instance As MainForm
    ' UR instance that handles connection to the robot
    Private ReadOnly _ur As UR = New UR()

    Public Shared Property Instance As MainForm
        Get
            Return _Instance
        End Get
        Private Set(value As MainForm)
            _Instance = value
        End Set
    End Property

#Region "Initialisation"
    Public Sub New()
        Instance = Me

        ' Catch all unhandled exceptions
        AddHandler AppDomain.CurrentDomain.UnhandledException, AddressOf CurrentDomain_UnhandledException
        AddHandler Application.ThreadException, AddressOf Application_ThreadException

        ' Load configuration file
        Config.Load()

        InitializeComponent()

        ' Add nodes in left menu and instanciate associated control 
        AddNode(New ConnectControl(_ur))
        AddNode(New PrimaryInterfaceControl(_ur))
        AddNode(New ProgramControl(_ur))
        AddNode(New RtdeControl(_ur))
        AddNode(New DashboardControl(_ur))
        AddNode(New InterpreterModeControl(_ur))
        AddNode(New XmlRpcControl(_ur))
        AddNode(New SocketControl(_ur))
        AddNode(New SftpControl(_ur))
        AddNode(New SshControl(_ur))
        AddNode(New ToolsControl())
        AddNode(New ArchiveControl(_ur))
        AddNode(New LicenseControl())
        AddNode(New LogControl())

        AddHandler _ur.InternalErrorOccured, AddressOf _ur_InternalErrorOccured

        ' Select first node at startup
        SelectNode(leftTreeView.Nodes(0))

    End Sub


    Private Sub AddNode(control As IUserControl)
        Dim node = leftTreeView.Nodes.Add(control.Title)
        node.Tag = control
    End Sub
#End Region

#Region "Unhandled exception"
    Private Sub Application_ThreadException(sender As Object, e As Threading.ThreadExceptionEventArgs)
        CatchApplicationException(e.Exception)
    End Sub

    Private Sub CurrentDomain_UnhandledException(sender As Object, e As UnhandledExceptionEventArgs)
        CatchApplicationException(TryCast(e.ExceptionObject, Exception))
    End Sub

    Private Sub CatchApplicationException(e As Exception)
        If InvokeRequired Then
            Invoke(New Action(Sub() CatchApplicationException(e)))
            Return
        End If

        MessageBox.Show(e?.Message, "An error occurred", MessageBoxButtons.OK, MessageBoxIcon.Error)
    End Sub
#End Region

#Region "User interaction"
    Private Sub leftTreeView_NodeMouseClick(sender As Object, e As TreeNodeMouseClickEventArgs)
        SelectNode(e.Node)
    End Sub

    Friend Function SelectNode(Of T As {Class, IUserControl})() As T
        Dim node = Instance.leftTreeView.Nodes.OfType(Of TreeNode)().FirstOrDefault(Function(n) TypeOf n.Tag Is T)
        Dim control = TryCast(node.Tag, T)
        SelectNode(node)
        Return control
    End Function

    ' Open right control associated to a node
    Friend Sub SelectNode(node As TreeNode)
        If leftTreeView.SelectedNode IsNot node Then leftTreeView.SelectedNode = node

        If node Is Nothing Then Return

        mainPanel.SuspendLayout()

        mainPanel.Controls.OfType(Of IUserControl)().FirstOrDefault()?.OnClose()

        mainPanel.Controls.Clear()

        Dim control = TryCast(node.Tag, Control)

        If control Is Nothing Then
            panelTitle.Text = ""
            Return
        End If

        panelTitle.Text = TryCast(node.Tag, IUserControl)?.Title

        mainPanel.Controls.Add(control)

        control.Dock = DockStyle.Fill

        mainPanel.ResumeLayout()

        mainPanel.Controls.OfType(Of IUserControl)().FirstOrDefault()?.OnOpen()
        mainPanel.Controls.OfType(Of IUserControl)().FirstOrDefault()?.PeriodicUpdate()
    End Sub

    ' Open browser to documentation page
    Private Sub lblLink_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs)
        Process.Start("https://underautomation.com/universal-robots/documentation?f")
    End Sub
#End Region

    ' Timer callback to refresh the control on the right and the left menu
    Private Sub tmrPeriodicUpdate_Tick(sender As Object, e As EventArgs)
        Try
            ' refresh control on the right
            mainPanel.Controls.OfType(Of IUserControl)().FirstOrDefault()?.PeriodicUpdate()
        Catch
        End Try

        lblStatus.Text = If(_ur.Enabled, $"Connected to {_ur.IP}", "Disconnected")

        Try
            ' Refresh left menu icons and colors
            For Each node As TreeNode In leftTreeView.Nodes
                Try
                    Dim ctrl = TryCast(node?.Tag, IUserControl)
                    If ctrl Is Nothing Then Continue For

                    Dim enabled = ctrl.FeatureEnabled

                    Dim key = If(enabled, "ON", "OFF")

                    If Not Equals(key, node.ImageKey) Then
                        node.ImageKey = key
                        node.SelectedImageKey = key
                        node.ForeColor = If(enabled, Color.Green, Color.DarkGray)
                    End If

                Catch
                End Try
            Next

        Catch
        End Try
    End Sub

    Private Sub titlePictureBox_MouseDoubleClick(sender As Object, e As MouseEventArgs)
        Generate(Me)
    End Sub

    Public Shared Sub Decompile(fullName As String)
        Dim control = Instance.SelectNode(Of ArchiveControl)()
        control?.Decompile(fullName)
    End Sub

    Public Shared Sub InvokeOnMainForm(action As [Delegate])
        If Instance.InvokeRequired Then
            Instance.Invoke(action)
            Return
        End If
        action?.DynamicInvoke()
    End Sub

#Region "Show errors"
    Public ReadOnly Errors As List(Of String) = New List(Of String)()

    Private Sub lblErrors_Click(sender As Object, e As EventArgs)
        Call SelectNode(Of LogControl)()
    End Sub

    ' Hide red label after some times
    Private Sub tmrError_Tick(sender As Object, e As EventArgs)
        tmrError.Enabled = False
        lblErrors.Text = ""
        lblErrors.Visible = False
    End Sub

    ' Callback when an error occured in SDK
    Private Sub _ur_InternalErrorOccured(sender As Object, e As InternalErrorEventArgs)
        Dim [error] = e.ToString()
        Errors.Add([error])
        ShowError([error])
    End Sub

    ' Display error
    Private Sub ShowError([error] As String)
        If InvokeRequired Then
            Invoke(New Action(Sub() ShowError([error])))
            Return
        End If

        If String.IsNullOrEmpty([error]) Then Return

        If [error].Length > 100 Then [error] = [error].Substring(0, 100) & " ..."

        lblErrors.Text = [error]
        lblErrors.Visible = True

        ' Show label some times
        tmrError.Enabled = True
    End Sub
#End Region
End Class
